### DISCUSSION TITLE
How to store/read DAGs from the database?

### DISCUSSION THREAD

**Discussion by joarobles** [apache/airflow#27746] [Q&A] [answered]
Hi there, I was checking the [docs](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/dagbag/index.html#airflow.models.dagbag.DagBag) and found out that DAGs can be read from the db setting the `read_dags_from_db` parameter but no further information is available.

> read_dags_from_db ([bool](https://docs.python.org/3/library/functions.html#bool)) – Read DAGs from DB if True is passed. If False DAGs are read from python files.

I was wondering:

1. What's the table schema like?
2. What are the pros/cons of using this approach instead of filesystem python files?
3. How could be this any different from generating dynamic DAGs from db query results? (as it's [discouraged in the docs](https://airflow.apache.org/docs/apache-airflow/stable/howto/dynamic-dag-generation.html))

**Comment by potiuk** [comment, marked as answer]
You can only read simplified representation this way (one that is used by scheduler and webserver to visualise DAGs). The format is internally serialized json - it's based on the DAG objects structure, not formalized in a schema.

You shoudl not use it unless you only want to know the structure of DAGs for reading and visualisation or making scheduling decisions (which you generally should not do). Ideally you should do it via API not as Python code if you wnt to use it in some external tools.

It has nothing to do with "dynamic DAG reading from DB query". The former is reading something that was created as the result of parsing, the latter should (not) be happening precisely during that parsing (so before the dags are serialized to DB).

**Reply by joarobles** [reply]
Thanks @potiuk, it not makes sense.

I was looking for alternatives on dynamic DAG generation (lots of them) because of the "magic loop" overload but I think I'll have to go back to generating .py files

**Comment by LucaSoato** [comment]
Hi, sorry for resurrecting this thread.
Is there a way to dynamically generate a dag that doesn't involve parsing py files?
For example, I'd like to programmatically take a simple function, serialize it and send it to the API/DB of Airflow from an external service. In that way I'd be able to generate as many dag I need, without making the parser check the files directly every 1~5 minutes.
This would also be more scalable, don't you think?

**Reply by potiuk** [reply]
Nope. There is no way this is directly opposing the way how airflow Architecture is currently. Some variations of it might be implemented in the future (and there are interesting ideas about it and likely they will be implemented in few minor releases) but the fact that for Airflow DAGs both structure and execution is result of parsing Python code today is something of a basic architectural decision of Airflow that a lot of other decisions base on.

**Reply by potiuk** [reply]
But if you have some ideas how to do it and create Airflow Improvemet Proposal on how to implement something like that - proposing that at Airflow devlist and writing proposal https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals are most welcome

**Reply by LucaSoato** [reply]
Thank you, Jarek!
I was inspired by the following quote from [this documentation page](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/dag-serialization.html):

> You can enable the source code to be stored in the database to make the Webserver completely independent of the DAG files.

I was checking how OpenMetadata has implemented its integration with Airflow. They literally write a JSON file that they use for dag configuration and a Python file with minimal code directly on the webserver, using [their plugin](https://github.com/open-metadata/OpenMetadata/tree/main/openmetadata-airflow-apis).
This makes the webserver (or any other component) stateful. Wouldn't it be possible to have a workaround that still allows an external service to create dags as simply as with an API call while keeping the PG database as the single source of truth? Which components in Airflow still depend on the fact that files generate the dags?
We also definitely want to contribute to Airflow if possible. If an Airflow Improvement Proposal can emerge from this, we'll be there!

**Reply by potiuk** [reply]
> You can enable the source code to be stored in the database to make the Webserver completely independent of the DAG files.

Yes. But only to display source code not to run it. There are two things stored there - source code of the DAG (but not its imports) to display it and serialized (json) form of the DAG structure to display it. This happens today (and it's not configurable - it's just how webserver works). But you still need the original code of DAG to execute it. One thing is to visualise DAG and another is to let worker execute it. Currently the basic assumption is that worker takes that defintion, parses it and executes the task it wants and there are PLENTY of things depending on it.

> dag configuration and a Python file with minimal code directly on the webserver, using [their plugin](https://github.com/open-metadata/OpenMetadata/tree/main/openmetadata-airflow-apis).

Yes. this is PRECISELY why airflow uses Python code as their definition. Anyone can write plugins which has a little Python code + some declarative form of defining how the code should map it in Python objects when the Python code is parrsed. This is the beauty of Airflow - because thanks to it's imperative definition of tasks, you can do ANYTHING. you can make any opinionated plugin like that bycause Airlfow Python parsing is super flexible. This way we do not have to define a declarative way how to describe a DAG. Instead there might be (and are) 100s of specialized ways how people do declartive definition on a subset of DAGs that implement specific cases. And that's exactly what we wanted to achieve. Contrary to popular belief it's easy to generate Python + configuration and by Airlfow doing parsing.

You have it backwards, really. The whole idea and basic premise of Airflow is that Python is the "common" way how to define and execute DAGs - both of them.

This is "Airflow as a Platform".

> Which components in Airflow still depend on the fact that files generate the dags?

DAG file Processor that creates the serialized definition of DAG from file and worker that executes thm. Worker is crucial because it does a lot - inclduding possibilty of dynamically applying Dynamic Task mapping from values returned by other tasks and yes - you guessed it - it does it in imperative code by running python code to map the values it receives from other tasks into actual tasks to execute. So the imperative definition of structure  of DAG (which is fixed once parsed) is not even nearly enough to actually get the dynamically mapped tasks running - code form DAG needs to be executed by the worker to pull the XCom and process it - sometimes also in a lazy way.

> We also definitely want to contribute to Airflow if possible. If an Airflow Improvement Proposal can emerge from this, we'll be there!

To be honest if you do not know which components need Python, then I strongly suggest to  dig deeper and do a lot of reading of the code documentation and past discussions. For one It took me about 6 months when I started to even grasp some of hte internals of Airflow and if you do not know which components depend on Python code, and want to cbange some absolutely basic assumptions of Airlfow - then I suggest you start small. 

Pick a good first issue (not in core area) and start contributing - loook through past discussions in devlist (you will find linnks in "community" section of Airflow, then gradually start picking more involved tasks and careully follow reviews and suggestions. Subscribe to devlist - read past discussions as much as you can and listen to discussions, then start to participate if you already grasp the concepts and basic assumptions. Look at the past AIPs and those that are currently discussed and try to understayd what they propose - maybe do some comments and reviews on thos. 

Proposing fundamental changes is maybe possible in your year 2 or 3 of Airlfow adventure if you spend significant time of yours to understand (that would be my estimate).

But of course, maybe you can do it faster, just make sure to show that you've done the homework. Discussinng half-baked ideas without knowing the consequences of it, especially when you want to challenge basic assumptions might lead to two things: you will know very well what you do and you will lead the change or you will start challenging things and asking for things that are pretty basic and people will not want to spend time explaining basics
